src/tools/rbd_mirror/group_replayer/BootstrapRequest.cc

```diff
@@ -31,1268 +36,256 @@ namespace group_replayer {
-    finish(r);
-    return;
-  }
-
-  m_local_mirror_group.state = cls::rbd::MIRROR_GROUP_STATE_DISABLED;
-  if (r != -ENOENT && (m_local_mirror_group.global_group_id == m_global_group_id) &&
-      (m_remote_mirror_group.state == cls::rbd::MIRROR_GROUP_STATE_ENABLED &&
-       m_remote_mirror_group_primary) && *m_resync_requested == false) {
-    create_local_mirror_group();
-  } else {
-    remove_local_group();
-  }
-}
-
-template <typename I>
-void BootstrapRequest<I>::remove_local_group() {
-  dout(10) << m_group_name << " " << *m_local_group_id << dendl;
-
-  ceph_assert(!m_local_group_id->empty());
-  ceph_assert(!m_group_name.empty());
-
-  librados::ObjectWriteOperation op;
-  op.remove();
-  auto comp = create_rados_callback<
-      BootstrapRequest<I>,
-      &BootstrapRequest<I>::handle_remove_local_group>(this);
-
-  int r = m_local_io_ctx.aio_operate(
-      librbd::util::group_header_name(*m_local_group_id), comp, &op);
-  ceph_assert(r == 0);
-  comp->release();
-}
-
-template <typename I>
-void BootstrapRequest<I>::handle_remove_local_group(int r) {
-  dout(10) << "r=" << r << dendl;
-
-  if (r < 0 && r != -ENOENT) {
-    derr << "error removing local group: " << cpp_strerror(r) << dendl;
+  } else if (r == -ENOENT) {
+    if (state_builder->remote_group_id.empty()) {
+      if (state_builder->local_group_id.empty()) {
+	// Neither group exists
+	m_local_group_removed = true; //FIXME
+	finish(0);
+	return;
+      } else if (state_builder->local_promotion_state ==
+                 librbd::mirror::PROMOTION_STATE_NON_PRIMARY){
+	remove_local_group();
+	return;
+      } else {
+	// Do not remove the group if the promotion state is orphan or unknown
+	finish(-ENOENT);
+	return;
+      }
+    } else if (state_builder->is_remote_primary()){
+      //FIXME: Check if state == disabled /delete the local group?
+      remove_local_group();
+      return;
+    }
+  } else if (r < 0) {
+    derr << "error preparing remote group: " << cpp_strerror(r)
+         << dendl;
     finish(r);
     return;
   }
 
-  remove_local_group_id();
-}
-
-template <typename I>
-void BootstrapRequest<I>::remove_local_group_id() {
-  dout(10) << dendl;
-
-  librados::ObjectWriteOperation op;
-  librbd::cls_client::group_dir_remove(&op, m_group_name, *m_local_group_id);
-
-  auto comp = create_rados_callback<
-      BootstrapRequest<I>,
-      &BootstrapRequest<I>::handle_remove_local_group_id>(this);
-
-  int r = m_local_io_ctx.aio_operate(RBD_GROUP_DIRECTORY, comp, &op);
-  ceph_assert(r == 0);
-  comp->release();
-}
-
-template <typename I>
-void BootstrapRequest<I>::handle_remove_local_group_id(int r) {
-  dout(10) << "r=" << r << dendl;
-
-  if (r < 0 && r != -ENOENT) {
-    derr << "error removing local group id: " << cpp_strerror(r) << dendl;
-    finish(r);
-    return;
+  if (!state_builder->is_remote_primary()) {
+//      if (state_builder->local_group_id.empty()) {
+    // local group does not exist and remote is not primary
```

Comment:
I guess the comment above is not valid for the line of code below because you don't check if `local_group_id` is empty? Please remove the comment above if not valid.

Reply:
This bit of code has been updated now.